design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
```

Comment:
why are these choices (fail-on etc) externalized? Shouldn't they be part of a Keycloak deployment choice vs who calls the healthcheck? Isn't the healthcheck caller dump?

Reply:
Could you please elaborate on this a bit more and give some examples?

Reply:
Let me answer your comment as I understand it...

Keycloak can be deployed in many different scenarios. In one deployment scenario, it might be just a thin layer to let users log in using different External IdP (Google, Facebook, Github etc.). In such case, if the External IdP you rely on is down, Keycloak also can't be considered as "healthy". 

Another example is a full blown SaaS approach. In this scenario you probably use a mix of different External IdPs, User Federation etc. In this example, if any of the External IdPs is down, there's nothing to worry about as user can log in using different methods. 

In the first case, you'd probably want to include the External IdP check in your readiness probe whereas in the latter not.

I believe the probe is very flexible this way.

Reply:
Actually, this made me think a bit, that it might be better to have some option on the server/realm level to configure what external services should be included in the health check, rather than making it the responsibility of the caller of the endpoint. You may for instance start out with using local username/passwords, but then for whatever reason decide to have a single external IdP instead. In this case Keycloak could actually be smart enough to already know this is an IdP it can't live without, so why then require someone to update the healthcheck URL? Also, how would you update the healthcheck URL when using the operator?

Reply:
I think putting this into server configuration doesn't buy us anything. Whenever you'd like to change the behavior of the probe, you'd probably need to restart the server or cycle the Pods. 

The only option I find interesting is putting the probe configuration into the server runtime. This probably means we'd need to put the configuration parameters somewhere into the database and provide some UI screen with the options. 

But then, the problem appears - the Admin Console is built around the Realm concept. The Readiness/Liveness probe is something that belongs to the server itself, not a particular Realm. So where to put a configuration screen for it? Also, is it worth the effort? I'd say not for the first iteration of this feature. In the future - perhaps?

Reply:
The problem here is that user federation and identity providers are a per-realm thing, while the probe is now an external thing to configure. Now if a realm admin adds a user federation provider, he would then have to communicate to the KC admin to update the probe. For a realm that has a single IdP at the realm level as I mentioned we can actually detect that and know it is important, without requiring any manual update to the probe.

Reply:
Another thing is that we now have user fed and idps at the global server level in the health check, while in reality one realm could be up and running just fine with a user fed provider being down, while another realm may rely on it for its service.

Reply:
A few comments below (here: https://github.com/keycloak/keycloak-community/pull/55#discussion_r496622654) we started a discussion whether or not External IdPs (and I also added User Federation to this context) belong to the Healthcheck API. Perhaps not? Perhaps they should be only an item in the monitoring dashboard?

Reply:
Together with @stianst we decided to remove the External IdPs and User Federation from Healthcheck API. See [this comment](https://github.com/keycloak/keycloak-community/pull/55#discussion_r498709614) for more explanation. 

Resolving this conversation.